sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala

```diff
@@ -90,19 +90,24 @@ trait FlatMapGroupsWithStateExecBase
 
   override def shortName: String = "flatMapGroupsWithState"
 
-  override def shouldRunAnotherBatch(newMetadata: OffsetSeqMetadata): Boolean = {
+  override def shouldRunAnotherBatch(newInputWatermark: Long): Boolean = {
     timeoutConf match {
       case ProcessingTimeTimeout =>
         true  // Always run batches to process timeouts
       case EventTimeTimeout =>
         // Process another non-data batch only if the watermark has changed in this executed plan
         eventTimeWatermarkForEviction.isDefined &&
-          newMetadata.batchWatermarkMs > eventTimeWatermarkForEviction.get
+          newInputWatermark > eventTimeWatermarkForEviction.get
       case _ =>
         false
     }
   }
 
+  // There is no guarantee that any of the column in the output is bound to the watermark. The
+  // user function is quite flexible. Hence Spark does not support the stateful operator(s) after
+  // (flat)MapGroupsWithState.
+  override def produceOutputWatermark(inputWatermarkMs: Long): Option[Long] = None
```

Comment:
I think this is ok for now, but what is the future plan to support emitting watermarks from flatMapGroupsWithState?  It is only an implementation limitation in SS that we cannot support stateful operators after a UDF.  I would imagine we can select the first event time column or let the use pick the event time column and use that for watermarking

Reply:
Spark has no idea about the value of the timestamp for the output. That said, either user should provide the logic to calculate the watermark (I don't think it will work), or the operator has to track the value for the event time (delayed by one batch). 

But even before we talk about this, there are lots of issues about the event time with typed API. During conversion between untyped <-> typed Spark loses the metadata in some cases. Furthermore, the output of flatMapGroupsWithState is also typed, which we need to convert back to untyped to apply further stateful operations, hence the risk of losing metadata. In addition, we don't have an ability of specifying new event time in the output for typed API.